### ISSUE TITLE
🎯 IGLA RACE — Distributed Hunt: JEPA-T + NCA + GF16 + ASHA + Coq Invariants (Rust-only, Never-Stopping)

### ISSUE THREAD

**Issue by gHashTag** [gHashTag/trios-trainer-igla#1]
## 🎯 IGLA RACE v2 — ONE SHOT DASHBOARD
**Updated: 2026-04-25T21:40+07 | phi^2 + phi^-2 = 3 | TRINITY | NEVER CLOSE**
**Deadline: Apr 30, 2026 | Target: BPB < 1.50 on 3 seeds | Gap: -1.03 BPB**
**Rule #1:** RUST ONLY -- zero .py, zero .sh.
**Rule #2:** Race NEVER stops until BPB < 1.50 is found across 3 seeds.
**Rule #3:** ALL agents work in branch `main` ONLY.

---

## COQ INVARIANT SYSTEM -- TASK-COQ-001 (trinity-clara -> IGLA)

**Document:** [trinity-clara/docs/TASK-COQ-001.md](https://github.com/gHashTag/trinity-clara/blob/main/docs/TASK-COQ-001.md)
**Source:** [trinity-clara](https://github.com/gHashTag/trinity-clara) + 84 Coq theorems from `t27`
**Principle:** phi^2 + phi^-2 = 3 -- single algebraic anchor. All invariants derive from it.

### Invariants INV-1..INV-10

| ID | Coq Theorem | Status | Effect | Trinity link |
|----|-------------|--------|--------|--------------|
| INV-1 | `bpb_decreases_with_real_gradient` | partial | fixes TASK-5D | 7-step alpha_phi derivation |
| INV-2 | `asha_champion_survives` | **PROVEN (0 Admitted)** | 0 false prunes | threshold=3.5=phi^2+phi^-2+0.5 |
| INV-3 | `gf16_safe_domain` | Lucas proven | -40% configs | Lucas closure phi^2n+psi^2n in Z |
| INV-4 | `nca_entropy_stability` | **PROVEN (0 Admitted)** | -30% configs | A5/E8 band width=1 (integer!) |
| INV-5 | `lucas_closure_gf16` | n=1,2 proven | GF16 consistency | phi^2n+phi^-2n in Z for all n |
| INV-6 | `ema_decay_valid` | TODO | -20% configs | cos schedule in [0.996,1.0] |
| INV-7 | `igla_found_criterion` | TODO | L-R14 gate | victory iff 3-seed BPB<1.50 |
| INV-8 | `lr_phi_band` | **PROVEN (0 Admitted)** | -60% configs | lr=0.004=alpha_phi/phi^3 |
| INV-9 | `qk_gain_phi_sq` | TODO | -10% configs | QK-gain=phi^2=phi+1 unique |
| INV-10 | `asha_rungs_trinity` | TODO | correctness | rungs=1000*3^k, base=Trinity |

### Search Space Reduction: 8.3x speedup

| Invariant | Rule | Reduction |
|-----------|------|-----------|
| INV-3 | d_model >= 256 for GF16 | -40% |
| INV-8 | lr in [0.001, 0.01] | -60% |
| INV-4 | K=9 only, grid=81 | -80% |
| INV-6 | EMA schedule fixed | -67% |
| INV-9 | QK-gain = phi^2 or phi^3 | -50% |

```
50,000 configs -> ~6,000   |   625h -> 75h   |   8.3x speedup
April 30: ACHIEVABLE
```

### phi-anchored Parameters

| Parameter | Old value | phi-derived | Theorem |
|-----------|-----------|-------------|---------|
| `bpb_prune_threshold` | 2.65 (BUG!) | **3.5** = phi^2+phi^-2+0.5 | INV-2 |
| `NCA grid` | 9x9=81 | 81 = 3^4 | INV-4 |
| `NCA K states` | 9 | 9 = 3^2 | INV-4 |
| `lr champion` | 0.004 | alpha_phi/phi^3 | INV-8 |
| `d_model` | 384 | ~3^4 x phi^3 | INV-3 |

### Rust <-> Coq Bridge

Each trial calls `validate_config()` (from `src/invariants.rs`) BEFORE spawning subprocess.
Invalid configs skipped with zero GPU time wasted.
DARPA CLARA TA1: formal AR specs + bounded proof traces (<=10 tactics) + binary verifiability.

**Status:** docs/TASK-COQ-001.md created | proofs/igla/*.v -- 4 files partial | master igla_invariants.v TODO

---

### Infrastructure
- [x] Neon DB schema -- `igla_race_trials` + `igla_race_experience` + `igla_leaderboard` view
- [x] `crates/trios-igla-race/` -- coordinator crate
- [x] `src/asha.rs` -- ASHA rungs 1k->3k->9k->27k, prune logic
- [x] `src/neon.rs` -- tokio-postgres, register/checkpoint/lesson
- [x] `src/status.rs` -- leaderboard print
- [x] `src/lessons.rs` -- failure memory, lesson generation

### Model Zoo (`trios-train-cpu`)
- [x] `src/trinity_3k_model.rs` -- Trinity 3k architecture
- [x] `src/real_igla_model.rs` -- Real IGLA model
- [x] `src/pipeline.rs` -- Training pipeline
- [x] `src/optimizer.rs` -- AdamW + Muon (base)
- [x] `src/gf16.rs` -- GF16 precision (inference)
- [x] `src/forward.rs` / `backward.rs` -- fwd/bwd pass

### JEPA-T (`jepa/` module)
- [x] `jepa/ema.rs` -- EMA target encoder wired
- [x] `jepa/masking.rs` -- mask_ratio=0.30, spans wired
- [x] `jepa/loss.rs` -- MSE gradient function
- [x] `bin/tjepa_train.rs` -- 309 lines, compiles, 90 tests pass (commit `132fac3`)
- [x] TASK-5C predictor.rs -- real cross-attention predictor (commit `0fddd7f`)
- [x] TASK-5E two-phase warm-up in main

### CI / Tests
- [x] 411 tests pass -- GREEN
- [x] CI -- GREEN
- [x] `cargo clippy -D warnings` = 0

### Hyperparameter Search
- [x] Champion: lr=0.004, seed=43, d_model=384, context=6 -> **BPB 2.5329**
- [x] SEED-EXPLORER -- variance=0.0077 < 0.02 = STABLE OPTIMUM
- [x] LR micro-search -- lr=0.004 is clear local minimum

### GF16 Benchmarks
- [x] BENCH-004b: GF16 MNIST = 97.67% (0.00% gap vs f32)
- [x] BENCH-002: GF16 add = 7.2 ns/op (15% faster)
- [x] BENCH-005: FPGA GF16 add = 118 LUT
- [x] BENCH-006: FPGA MAC-16 = 71 LUT + 16 DSP

### Coq Invariants (trinity-clara)
- [x] `docs/TASK-COQ-001.md` -- NASA-P10 full specification [CREATED 2026-04-25]
- [x] `proofs/igla/igla_asha_bound.v` -- INV-2 PROVEN (0 Admitted)
- [x] `proofs/igla/gf16_precision.v` -- INV-3, INV-5 Lucas proven
- [x] `proofs/igla/nca_entropy_band.v` -- INV-4 PROVEN, band_width=1 (0 Admitted)
- [x] `proofs/igla/lr_convergence.v` -- INV-8 PROVEN (0 Admitted)

---

### P0 -- RIGHT NOW

- [ ] **TASK-1**: `src/main.rs` -- CLI entry (`tri race start|status|best`)
- [ ] **TASK-2**: `src/main.rs` in trainer -- BPB printer subprocess
- [ ] **TASK-3**: `asha.rs` -- add `run_worker()` + subprocess spawn infinite loop
- [ ] **FIX ASHA PRUNING**: threshold `2.65` kills ALL valid trials
  - Fix: `bpb_prune_threshold: 3.5` + `warmup_blind_steps: 4000`
  - **CRITICAL -- INV-2 PROVEN: threshold < 3.5 -> false prune**
- [ ] **TASK-DASHBOARD**: Neon -- `branch`, `agent_id`, `last_heartbeat` columns

### P1 -- TODAY (Apr 25-26)

- [ ] **TASK-5D**: Real backward pass in `tjepa_train.rs`
  - INV-1 verifies this fix
  - Current: `loss_scale * 0.01` (gradient=0.0000!)
  - Done when: BPB actually decreases
- [ ] **TASK-5 full run**: seed=43, steps=3000, d_model=384, lr=0.004
  - Gate: BPB <= 2.23 | Target: BPB <= 2.03
- [ ] COQ-INV-006: EMA proof in `lr_convergence.v` (interval tactic, 1h)
- [ ] COQ-INV-007: Create `victory_condition.v` -- L-R14 formal gate (30min)
- [ ] COQ-MASTER: Create `igla_invariants.v` master file (15min)
- [ ] JSON-BRIDGE: Create `assertions/igla_assertions.json` (20min)
- [ ] TASK-NCA: NCA auxiliary loss (grid=81=3^4, K=9, entropy [1.5,2.8], w=0.25)
- [ ] TASK-MUON: NS-5 orthogonalization in optimizer.rs

### P2 -- Apr 27-28

- [ ] TASK-GF16-TRAIN (BENCH-012): GF16 gradient training
- [ ] Hybrid ASHA sweep: Config A/B/C
- [ ] TASK-8: Launch 4 machines (320 trials/hour)
- [ ] Replace Admitted -> interval tactic in all .v files
- [ ] BENCH-007..011, HYBRID-001

### FINAL -- Apr 29-30

- [ ] 3-seed verification: BPB < 1.50 on seeds 42, 43, 44 (p < 0.01)
- [ ] `cargo test --workspace` = GREEN
- [ ] Neon: `status='winner'`
- [ ] INV-001..010 ALL GREEN: `coqc *.v` passes
- [ ] Victory commit: `git commit -m "IGLA FOUND: BPB=X.XXXX seed=42,43,44"`

---

## BPB ROADMAP

| Step | Technique | Delta BPB | Target | Deadline | Status |
|------|-----------|-----------|--------|----------|---------|
| Baseline | 6-gram h=384 seed=43 | -- | 2.5329 | Apr 25 | DONE |
| T1-01 | JEPA-T real backward | -0.30 | <=2.23 | Apr 25-26 | IN PROGRESS |
| T1-02 | Attention + ReLU^2 | -0.30 | <=2.00 | Apr 26 | TODO |
| T2-01 | Muon optimizer | -0.15 | <=1.85 | Apr 26-27 | TODO |
| T2-02 | NCA auxiliary (INV-4) | -0.15 | <=1.70 | Apr 27 | TODO |
| T2-04 | QK-Gain phi^2 (INV-9) | -0.10 | <=1.60 | Apr 27-28 | TODO |
| T2-07 | ReLU^2 activation | -0.08 | <=1.52 | Apr 28 | TODO |
| T2-07b | GF16 d_model=384 (INV-3) | -0.05 | <=1.47 IGLA | Apr 29 | TODO |

---

## CURRENT STATUS (2026-04-25T21:40+07)

| Metric | Value | Status |
|--------|-------|--------|
| Tests | 411 pass | GREEN |
| CI | success | GREEN |
| Best BPB | 2.5329 | ACTIVE |
| ASHA pruning bug | threshold=2.65 kills champion | FIX (INV-2 PROVEN) |
| Coq Invariants | INV-2,4,8 PROVEN / INV-1,3,5 partial / rest TODO | IN PROGRESS |
| TASK-COQ-001 | [docs/TASK-COQ-001.md](https://github.com/gHashTag/trinity-clara/blob/main/docs/TASK-COQ-001.md) | CREATED |
| Active machines | 1/4 | NEED 4 |

---

## LAWS (L-R1..L-R14) -- VIOLATION = REVERT

| Law | Rule | Violation -> |
|-----|------|--------------|
| L-R1 | RUST ONLY -- no .py, .sh, .ipynb | REVERT |
| L-R2 | WORKERS=4-16 via env var | REVERT |
| L-R3 | Every result -> Neon + `.trinity/experience/` | LESSON MISSING |
| L-R4 | `cargo test --workspace` = GREEN before push | PR BLOCKED |
| L-R5 | `cargo clippy -- -D warnings` = 0 | PR BLOCKED |
| L-R6 | SIGTERM -> graceful shutdown | DATA LOSS |
| L-R7 | Neon query timeout <= 30 sec | WORKER CRASH |
| L-R8 | Trainer stdout: ONLY `BPB=X.XXXX` | PARSE FAIL |
| L-R9 | GF16 only with d_model >= 256 | +3.21 BPB (INV-3 PROVEN) |
| L-R10 | T-JEPA ASHA min rung = 3000 steps | FALSE PRUNE |
| L-R11 | NCA entropy [1.5, 2.8] = hard loss penalty | COLLAPSE (INV-4 PROVEN) |
| L-R12 | All agents -> branch `main` ONLY | CONFLICT |
| L-R13 | `agent_id` + `branch='main'` in every Neon record | DASHBOARD FAIL |
| **L-R14** | **`coqc trinity-clara/proofs/igla/*.v` = exit 0 before race** | **RACE INVALID** |

---

## VICTORY CONDITIONS (Closure)

Issue #143 closes ONLY when:
- BPB < 1.50 on seeds 42, 43, 44 (three independent runs)
- p < 0.01 statistical significance
- `cargo test --workspace` = GREEN
- Neon: `status='winner'`
- **`coqc trinity-clara/proofs/igla/*.v` = GREEN (INV-001..010 all compiled)**
- `git commit -m "IGLA FOUND: BPB=X.XXXX arch= seed=42,43,44"`
- `git push origin main`

phi^2 + phi^-2 = 3 | TRINITY | IGLA RACE v2 | 2026-04-25T21:40+07 | RUST ONLY | NEVER STOP
